src/lib.rs

```diff
@@ -182,7 +208,20 @@ where
 
         for pattern_token in pattern_tokens {
             for token in self.reverse_map.keys() {
-                let score = jaro_winkler(token, &pattern_token);
+                let score = if self.option.levenshtein {
+                    let len = max(token.len(), pattern_token.len()) as f64;
+                    // calculate k (based on the threshold) to bound the Levenshtein distance
+                    let k = ((1.0 - self.option.threshold) * len).ceil() as u32;
+                    // levenshtein_simd_k only works on ASCII byte slices, so the token strings
+                    // are directly treated as byte slices
+                    match levenshtein_simd_k(token.as_bytes(), pattern_token.as_bytes(), k) {
```

Comment:
Can we directly `unwrap()` on it? It seems not possible to have both token and pattern_token empty.

Reply:
https://github.com/andylokandy/simsearch-rs/blob/19cd803de7b5c8098687b3fe36fd353f6672d9de/src/lib.rs#L300

Reply:
Oh no, the reason why it returns an `Option` is because the tokens may not match within the threshold `k`. This is efficient, because we don't have to figure out the score if it is going to be greater than the threshold, and we don't need the exact value of the score anyways if it is less than the threshold. Therefore, cannot directly unwrap it. The length check was just to be safe, as I did not want to check everywhere to ensure that the length will always be nonzero.

Reply:
I got you, but why score the inputs under the threshold infinitely low? It will get other tokens ignored.

Reply:
If two tokens do not match within `k` edits, then we score them infinitely low, so they are automatically ignored because they don't contain enough matching characters. This `k` is calculated so that if two tokens meet the required threshold, then they *must* have an edit distance less than `k`. The threshold is a lower bound of the number of edits as a percentage, while `k` needs to be an upper bound on the number of edits. Therefore, we need to convert the threshold to `k` before calling my library's function, and then convert the output (number of edits) to a percentage that measures similarity.

Reply:
As you can see from here:

https://github.com/andylokandy/simsearch-rs/blob/19cd803de7b5c8098687b3fe36fd353f6672d9de/src/lib.rs#L233-L237

The final score is computed from the sum of all combinations of source tokens and patterns tokens. For example, if we perform a search of "old" over "the old man and the sea", we will get an infinitely low score on the first iteration which compares "old" with "the".

Reply:
Right, but isn't there a check here:
https://github.com/andylokandy/simsearch-rs/pull/7/files/19cd803de7b5c8098687b3fe36fd353f6672d9de#diff-b4aea3e418ccdb71239b96952d9cddb6R225-R227
That causes the infinitely low score to be intentionally ignored when summing up scores later on?

Reply:
Isn't there a check afterwards:
```suggestion
                if score > self.option.threshold {
                    token_scores.insert(token, score);
                }
```
that will eliminate the infinitely negative scores? Then, the score of the token "the" will never be included in the summation.

Reply:
Oh, yes... I missed it totally.